File: packages/python-sdk/src/vocab_bloom_hub/cache.py
"""Conditional requests for GET answers.

The public API sends an ``ETag`` with every successful GET and answers
``304 Not Modified`` to a matching ``If-None-Match``. The cache keeps the
last body per URL and hands it back on a 304, so a repeated read costs a
round trip without a payload. In memory, per client instance, bounded by
``max_entries`` (least recently used out).
"""

from __future__ import annotations

from collections import OrderedDict
from dataclasses import dataclass
from typing import Any, Protocol


@dataclass(frozen=True)
class CacheEntry:
    etag: str
    body: Any


class ResponseCache(Protocol):
    def get(self, url: str) -> CacheEntry | None: ...

    def set(self, url: str, entry: CacheEntry) -> None: ...


class MemoryCache:
    def __init__(self, max_entries: int = 500) -> None:
        self._entries: OrderedDict[str, CacheEntry] = OrderedDict()
        self._max_entries = max_entries

    def get(self, url: str) -> CacheEntry | None:
        entry = self._entries.get(url)
        if entry is not None:
            self._entries.move_to_end(url)
        return entry

    def set(self, url: str, entry: CacheEntry) -> None:
        self._entries[url] = entry
        self._entries.move_to_end(url)
        while len(self._entries) > self._max_entries:
            self._entries.popitem(last=False)

    def __len__(self) -> int:
        return len(self._entries)

File: packages/npm-sdk/src/cache.ts
/**
 * Conditional requests for GET answers: the public API sends an ETag with
 * every successful GET and answers `304 Not Modified` to a matching
 * `If-None-Match`. The cache keeps the last body per URL and hands it back
 * on a 304, so a repeated read costs a round trip without a payload.
 * In-memory, per client instance, bounded by `maxEntries` (oldest out).
 */
export type CacheEntryT = { etag: string; body: unknown };

export interface ResponseCache {
  get(url: string): CacheEntryT | undefined;
  set(url: string, entry: CacheEntryT): void;
}

export class MemoryCache implements ResponseCache {
  private readonly entries = new Map<string, CacheEntryT>();

  constructor(private readonly maxEntries = 500) {}

  get(url: string): CacheEntryT | undefined {
    const entry = this.entries.get(url);
    if (entry) {
      // refresh the insertion order: the least recently used goes first
      this.entries.delete(url);
      this.entries.set(url, entry);
    }
    return entry;
  }

  set(url: string, entry: CacheEntryT): void {
    this.entries.delete(url);
    this.entries.set(url, entry);
    while (this.entries.size > this.maxEntries) {
      const oldest = this.entries.keys().next().value;
      if (oldest === undefined) break;
      this.entries.delete(oldest);
    }
  }

  get size(): number {
    return this.entries.size;
  }
}
